### PR TITLE
Assets mapper

### PR DESCRIPTION
**Pull request by Kicer86**

### CODE REVIEW

**Comment by Kicer86** [comment]
Here is an example of config.json file which needs to be placed into resources directory:


{
    "files":
    {
        "clean water": "env/cleanwater.png"
    }
}

File: assets/ConfigReader.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        struct Config: IConfig
+        {
+            Config(const std::map<std::string, std::string>& m)
+                : m_map(m)
+            {
+
+            }
+
+            std::string file_path(const std::string& asset_name) const
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: Mixing methods naming convention. Please use camelCase.

File: assets/ConfigReader.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                const auto it = m_map.find(asset_name);
+                return it == m_map.cend()? std::string(): it->second;
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: One-line if unreadable. Please use more spaces.

File: assets/ConfigReader.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
+/*
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: Use camelCase in entire file.

File: assets/ConfigReader.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            std::string file_path(const std::string& asset_name) const
+            {
+                const auto it = m_map.find(asset_name);
+                return it == m_map.cend()? std::string(): it->second;
+            }
+
+            const std::map<std::string, std::string> m_map;
+        };
+
+        void throw_on_false(bool b)
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: This helper function is ugly.

File: assets/ConfigReader.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        Json::CharReaderBuilder builder;
+        Json::Value json_config;
+
+        std::map<std::string, std::string> entries_mapping;
+        JSONCPP_STRING errs;
+        try
+        {
+            throw_on_false( parseFromStream(builder, config, &json_config, &errs) );
+
+            const Json::Value mapping_entry = json_config["files"];
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: Use constant for "files" string. (May be used in future).

File: assets/ConfigReader.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+            for(const auto& entry: entries)
+            {
+                const Json::Value& value = mapping_entry[entry];
+                throw_on_false( value.isString() );
+
+                entries_mapping[entry] = value.asString();
+            }
+        }
+        catch(const std::exception &)
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: Try to catch all possible exceptions.

File: assets/ConfigReader.cpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    ConfigReader::ConfigReader()
+    {
+
+    }
+
+
+    std::unique_ptr<IConfig> ConfigReader::load()
+    {
+        std::ifstream config;
+        config.open("game-assets/config.json");
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: Use constant.

File: assets/ConfigReader.cpp (hunk 4 above)
Comment (Kicer86) [reply assets/ConfigReader.cpp]: ?!?!?!?!?!?!!!!!!!?!?!?!?!?!?!oneoneone!!!

File: assets/ConfigReader.cpp (hunk 6 above)
Comment (Kicer86) [reply assets/ConfigReader.cpp]: like which?

File: assets/ConfigReader.cpp (hunk 4 above)
Comment (ThudPoland) [reply assets/ConfigReader.cpp]: This helper function doesn't describe why throwing exception is needed.
We should instead create unit tests.

Comment (Kicer86) [reply assets/ConfigReader.cpp]: it will blow when file structure is corrupted

File: assets/ConfigReader.cpp (hunk 6 above)
Comment (ThudPoland) [reply assets/ConfigReader.cpp]: Like proper exception for each case - if something is not in table, throw - out of range for example.

File: assets/ConfigReader.cpp [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            std::string filePath(const std::string& asset_name) const
+            {
+                const auto it = m_map.find(asset_name);
+
+                return it == m_map.cend()?
+                             std::string():
+                             it->second;
+            }
+
+            const std::map<std::string, std::string> m_map;
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: Naming convention.

File: assets/ConfigReader.cpp [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+    }
+
+
+    std::unique_ptr<IConfig> ConfigReader::load()
+    {
+        std::ifstream config;
+        config.open(jsonMappingsFilePath);
+
+        Json::CharReaderBuilder builder;
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: Use camelCase rather than snake_case.

File: assets/ConfigReader.cpp [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+#include <json/json.h>
+
+#include "IConfig.hpp"
+
+namespace Assets
+{
+    namespace
+    {
+        const char* const jsonMappingsFilePath = "game-assets/config.json";
+        const char* const jsonFilesMappingGroup = "files";
```
Comment (ThudPoland) [inline assets/ConfigReader.cpp]: Use rather std::string.

**Pull request merged**